src/show.jl

```diff
@@ -126,85 +119,107 @@ Configurable option to control whether emoji icons (`true`) or a plain-text anno
-               "[?]"
+        show_tree(io, obj)
     end
 end
+
+_tree_icon(::Type{Attribute}) = SHOW_TREE_ICONS[] ? "🏷️" : "[A]"
+_tree_icon(::Type{Group})     = SHOW_TREE_ICONS[] ? "📂" : "[G]"
+_tree_icon(::Type{Dataset})   = SHOW_TREE_ICONS[] ? "🔢" : "[D]"
+_tree_icon(::Type{Datatype})  = SHOW_TREE_ICONS[] ? "📄" : "[T]"
+_tree_icon(::Type{File})      = SHOW_TREE_ICONS[] ? "🗂️" : "[F]"
+_tree_icon(::Type)            = SHOW_TREE_ICONS[] ? "❓" : "[?]"
+_tree_icon(obj) = _tree_icon(typeof(obj))
 _tree_icon(obj::Attributes) = _tree_icon(obj.parent)
 
 _tree_head(io::IO, obj) = print(io, _tree_icon(obj), " ", obj)
 _tree_head(io::IO, obj::Datatype) = print(io, _tree_icon(obj), " HDF5.Datatype: ", name(obj))
 
-function _tree_children(parent::Union{File, Group}, attributes::Bool)
-    names = keys(parent)
-    objs  = Union{Object, Attribute}[parent[n] for n in names]
-    if attributes
-        attrn = keys(HDF5.attributes(parent))
-        attro = Union{Object, Attribute}[HDF5.attributes(parent)[n] for n in attrn]
-        names = append!(attrn, names)
-        objs  = append!(attro, objs)
+_tree_count(parent::Union{File,Group}, attributes::Bool) =
+    length(parent) + (attributes ? length(HDF5.attributes(parent)) : 0)
+_tree_count(parent::Dataset, attributes::Bool) =
+    attributes ? length(HDF5.attributes(parent)) : 0
+_tree_count(parent::Attributes, _::Bool) = length(parent)
+_tree_count(parent::Union{Attribute,Datatype}, _::Bool) = 0
+
+function _show_tree(io::IO, obj::Union{File,Group,Dataset,Datatype,Attributes,Attribute}, indent::String="";
+                    attributes::Bool = true, depth::Int = 1)
+    isempty(indent) && _tree_head(io, obj)
+    isvalid(obj) || return
+
+    INDENT = "   "
+    PIPE   = "│  "
+    TEE    = "├─ "
+    ELBOW  = "└─ "
+
+    limit = get(io, :limit, false)::Bool
+    counter = 0
+    nchildren = _tree_count(obj, attributes)
+
+    @inline function childstr(io, n, more=" ")
+        print(io, "\n", indent, ELBOW * "(", n, more, n == 1 ? "child" : "children", ")")
     end
-    return (names, objs)
-end
-function _tree_children(parent::Dataset, attributes::Bool)
-    names = String[]
-    objs = Union{Object, Attribute}[parent[n] for n in names]
+    @inline function depth_check()
+        counter += 1
+        if limit && counter > max(2, SHOW_TREE_MAX_CHILDREN[] ÷ depth)
+            childstr(io, nchildren - counter + 1, " more ")
+            return true
+        end
+        return false
+    end
+
+    if limit && nchildren > 0 && depth > SHOW_TREE_MAX_DEPTH[]
+        childstr(io, nchildren)
+        return nothing
+    end
+
     if attributes
-        attrn = keys(HDF5.attributes(parent))
-        attro = Union{Object, Attribute}[HDF5.attributes(parent)[n] for n in attrn]
-        names = append!(attrn, names)
-        objs  = append!(attro, objs)
+        obj′ = obj isa Attributes ? obj.parent : obj
+        h5a_iterate(obj′, H5_INDEX_NAME, H5_ITER_INC) do _, cname, _, _
```

Comment:
Is it worth it to set up this custom call-back instead of directly iterating over the attributes, which would be more readable and simpler to maintain?

Reply:
Do you mean a custom `h5a_iterate(::Attributes, ...)`? I'd thought about that, but didn't since it could also make sense to instead define `Base.cconvert(::Type{hid_t}, a::Attributes) = cconvert(hid_t, a.parent)`, but I hadn't thought/dug through the implications of enabling that yet.

Reply:
I guess we don't have a way to directly iterate over Attributes as in https://github.com/JuliaIO/HDF5.jl/blob/master/src/HDF5.jl#L956

Reply:
Oh, the other thing about the iteration is that at least for attributes, we don't ever actually have to open the objects — just having the name is sufficient. The [hypothetical] `Base.iterate` method returns an object, so that's extra allocations/libhdf5 overhead we can avoid by using `h5a_iterate` instead.